recipes/boost/all/conanfile.py

```diff
@@ -852,102 +911,126 @@ def _is_versioned_layout(self):
-                if os.path.splitext(real_lib_name)[0].split("-")[0].endswith(alib):
-                    ordered_libs[pos].append(real_lib_name)
-                    break
-            else:
-                # self.output.info("Missing in order: %s" % real_lib_name)
-                if "_exec_monitor" not in real_lib_name:  # https://github.com/bincrafters/community/issues/94
-                    missing_order_info.append(real_lib_name)  # Assume they do not depend on other
+                self.cpp_info.components["headers"].includedirs = [os.path.join(self.package_folder, "include", boost_version_tag)]
 
-        # Flat the list and append the missing order
-        self.cpp_info.libs = [item for sublist in ordered_libs
-                                      for item in sublist if sublist] + missing_order_info
+        self.cpp_info.components["headers"].libs = []
+        self.cpp_info.components["headers"].names["cmake_find_package"] = "boost"
+        self.cpp_info.components["headers"].names["cmake_find_package_multi"] = "boost"
 
-        if self.options.without_test:  # remove boost_unit_test_framework
-            self.cpp_info.libs = [lib for lib in self.cpp_info.libs if "unit_test" not in lib]
+        libformatdata = {}
+        if not self.options.without_python:
+            pyversion = tools.Version(self._python_version)
+            libformatdata["py_major"] = pyversion.major
+            libformatdata["py_minor"] = pyversion.minor
+
+        modules_seen = set()
+        detected_libraries = set(tools.collect_libs(self))
+        used_libraries = set()
+        for module in self._iter_modules():
+            if self.options.get_safe("without_{}".format(module), False) or not all(d in modules_seen for d in self._dependencies["dependencies"][module]):
+                continue
+            module_libraries = [lib.format(**libformatdata) for lib in self._dependencies["libs"][module]]
+            if all(l in detected_libraries for l in module_libraries):
+                modules_seen.add(module)
+
+                used_libraries = used_libraries.union(module_libraries)
+                self.cpp_info.components[module].libs = module_libraries
+
+                self.cpp_info.components[module].requires = self._dependencies["dependencies"][module] + ["headers"]
+                self.cpp_info.components[module].names["cmake_find_package"] = module
+                self.cpp_info.components[module].names["cmake_find_package_multi"] = module
+
+        if used_libraries != detected_libraries:
+            non_used = detected_libraries.difference(used_libraries)
+            if non_used:
+                raise ConanInvalidConfiguration("These libraries were not used in conan components: {}".format(non_used))
+
+            non_existing = used_libraries.difference(detected_libraries)
+            if non_existing:
+                raise ConanInvalidConfiguration("These libraries were used, but not built: {}".format(non_existing))
```

Comment:
We cannot raise `ConanInvalidConfiguration` here ATM. Main problem right now is that CCI captures this exception and mark it as a successful build, that's the reason why some it reports 131 packages built, but some of them didn't generated a package (check [`boost/1.72.0` package-id: `3a0121f23dce7b5f89659d27665f62cec98f40c5`](https://c3i.jfrog.io/c3i/misc/logs/pr/2097/23/boost/1.72.0/3a0121f23dce7b5f89659d27665f62cec98f40c5/)).

I'll add a development to CCI to propagate this error and fail the job. If the configuration cannot be built it should raise from the `configure` method. Given that, this job will fail with this exception here.

Anyway, this check probably shouldn't raise a `ConanInvalidConfiguration` but a `ConanException`, at least this is not the meaning and purpose of the `InvalidConfiguration` exception.

Reply:
@madebr I think we need to move this somewhere. we usually don't raise exceptions from the `package_info` as it's too late - like a final study when the package is already built/downloaded. this basically means package built is broken and cannot be used at all - we have to prevent that.
probably, right now we should explicitly encode some default options (e.g. `without_mpi` and `without_graph_parallel`) to avoid failures.

Reply:
This is more of a self-test.
Maybe they should become asserts, or a ConanException?

Reply:
probably, assert is better for self-test here then. ConanException is not something idiomatically correct here, as well as not well supported by the current CCI infrastructure